### ISSUE TITLE
AWS Error when Running XCode 7 UITests with fastlane/snapshot

### ISSUE THREAD

**Issue by CareerFairPlus** [aws/aws-sdk-ios#381]
**Background**
My company manages a few dozen iOS apps and, for the past year or so we have been using the [Fastlane](https://fastlane.tools/) toolchain to handle building and deploying updates and new apps.  Recently, the [Snapshot](https://github.com/fastlane/fastlane/tree/master/snapshot) tool in Fastlane was updated to use the new UITest framework in XCode 7 rather than the old UIAutomation framework.

Our app also uses the AWS SDK (probably an old version) to download data from an S3 bucket.  We do not currently upload any data to S3, so this is a read-only connection.

I have followed the [Snapshot setup guide](https://github.com/fastlane/fastlane/tree/master/snapshot#quick-start) for creating the required UITests and integrating with the SnapshotHelper.swift file that is used to setup whatever Snapshot needs to operate effectively.

**My UITest file is as follows:**
```
#import <XCTest/XCTest.h>

#import "CF_PlusUITests_Snapshot-Swift.h"

@interface CF_PlusUITests_Snapshot : XCTestCase

@end

@implementation CF_PlusUITests_Snapshot

- (void)setUp {
    [super setUp];
    
    // Put setup code here. This method is called before the invocation of each test method in the class.
    
    // In UI tests it is usually best to stop immediately when a failure occurs.
    self.continueAfterFailure = NO;
    // UI tests must launch the application that they test. Doing this in setup will make sure it happens for each test method.
    XCUIApplication *app = [[XCUIApplication alloc] init];
    
    // Setup snapshot here
    [Snapshot setupSnapshot:app];
    
    [app launch];
    
    // In UI tests it’s important to set the initial state - such as interface orientation - required for your tests before they run. The setUp method is a good place to do this.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (void)testExample {
    // Use recording to get started writing UI tests.
    // Use XCTAssert and related functions to verify your tests produce the correct results.
    
    /* Basic UI Test for Screenshots */
    XCUIApplication *app = [[XCUIApplication alloc] init];
    
    sleep(10);
    
    XCUIElement *firstChild = [[[app tables] childrenMatchingType:XCUIElementTypeCell] elementBoundByIndex:0];
    if (firstChild) {
        [firstChild tap];
    } else {
        NSLog(@"There is no first child.");
    }
    
    sleep(5);
}

@end
```

**The Problem**
Running these tests _without_ the `[Snapshot setupSnapshot:app];` line has no problems.  The tests execute normally and end successfully.

After adding the `[Snapshot setupSnapshot:app];` line, running the tests results in the app throwing the following error when attempting to access our AWS bucket:

```
Error Domain=com.amazonaws.AWSGeneralErrorDomain Code=3 "The request signature we calculated does not match the signature you provided. Check your key and signing method." UserInfo={NSLocalizedDescription=The request signature we calculated does not match the signature you provided. Check your key and signing method.}
```

I've looked through the `setupSnapshot` function and see that it appears to:

1. Set the device language.
2. Set the device locale.
3. Set some launch arguments (`["-FASTLANE_SNAPSHOT", "YES", "-ui_testing"]`).

I'm not sure how or why any of these would have an adverse affect on our AWS implementation, since we're currently using statically-defined credentials.

For additional information, I did create an issue for this on the Snapshot side ([link](https://github.com/fastlane/fastlane/issues/4366)).

Any ideas - crazy or not - that anyone has about what might be causing this, ways to solve it, or work-arounds would be greatly appreciated!  Thank you all in advance.

**Comment by yosuke-matsuda** [comment]
Can you share the code snippet that you are using to set up the AWS Mobile SDK?

**Comment by cdunkel** [comment]
@yosuke-matsuda - 

Here is the relevant code from my **AppManifest.m**:
```
#import <AWSiOSSDKv2/AWSCore.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {

    NSString* accessKey = @"<access key>";
    NSString* secretKey = @"<secret key>";
    
    AWSStaticCredentialsProvider *credentialsProvider = [AWSStaticCredentialsProvider credentialsWithAccessKey:accessKey secretKey:secretKey];
    AWSServiceConfiguration *configuration = [AWSServiceConfiguration configurationWithRegion:AWSRegionUSEast1 credentialsProvider:credentialsProvider];
    [AWSServiceManager defaultServiceManager].defaultServiceConfiguration = configuration;

}

@end
```

**Comment by CareerFairPlus** [comment]
@yosuke-matsuda - 

Here is the relevant code from my **AppManifest.m**:
```
#import <AWSiOSSDKv2/AWSCore.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {

    NSString* accessKey = @"<access key>";
    NSString* secretKey = @"<secret key>";
    
    AWSStaticCredentialsProvider *credentialsProvider = [AWSStaticCredentialsProvider credentialsWithAccessKey:accessKey secretKey:secretKey];
    AWSServiceConfiguration *configuration = [AWSServiceConfiguration configurationWithRegion:AWSRegionUSEast1 credentialsProvider:credentialsProvider];
    [AWSServiceManager defaultServiceManager].defaultServiceConfiguration = configuration;

}

@end
```

**Comment by yosuke-matsuda** [comment]
Your SDK seems to be outdated. You should use the latest version. Also, you should turn on the [verbose logging](http://docs.aws.amazon.com/mobile/sdkforios/developerguide/setup.html#logging) and take a look at the logs. It may give you some insights on why you are getting the error.

**Comment by CareerFairPlus** [comment]
Thank you for the note on verbose logging.  I will see what that gives me and post any follow-up questions here.

**Comment by yosuke-matsuda** [comment]
I cannot tell exactly what version of the SDK you are using, but you can take a look at the list of [Release Notes](https://aws.amazon.com/releasenotes/iOS) for the list of fixes for each version.

**Comment by CareerFairPlus** [comment]
@yosuke-matsuda I believe I found the issue with the verbose logs.

When the tests were failing, the user agent being reported by AWS was `user-agent:aws-sdk-iOS/2.0.4 iPhone-OS/9.3`.

When the tests passed, the user agent was `user-agent:aws-sdk-iOS/2.0.4 iPhone-OS/9.3 en_US`.

As part of the setup for the Snapshot utility we're using, it sets the device language and locale based on a text file.  Because this text file didn't exist the setup function just set the language to "".  Once I created the file and added a valid language my tests passed again.

Because this does not appear to be an issue with the AWS library, but rather with the how Snapshot deals with initial conditions, I will close this issue.

Thanks again for all the help!
